### ISSUE TITLE
Replace runtime_static_assert tests with compile tests.

### ISSUE THREAD

**Issue by allisonvacanti** [NVIDIA/cccl#807]
Some of the Thrust unit tests check that static assertions fire under particular situations. For example, [generate_const_iterators](https://github.com/NVIDIA/thrust/blob/81d683e42047913dcdcf7239df282806e2dc019c/testing/generate_const_iterators.cu) tests that `thrust::generate` and `thrust::fill` trigger static assertions when called with const iterators.

This is implemented by redefining the `THRUST_STATIC_ASSERT` macros to instead call a function that sets some state, which is verified in the unit test. This has the unfortunate side effect of making the `THRUST_STATIC_ASSERT` macro unusable in certain situations:

```
0 |   template <typename T>
1 |   struct Foo
2 |   {
3 |     THRUST_STATIC_ASSERT(sizeof(T) <= 4);
4 |   };
```

Adding such a usage of `THRUST_STATIC_ASSERT` to any header pulled in by these tests results in some inscrutable build failures. `nvcc` fails with:

```
foo.h(3): error: expected a type specifier
foo.h(3): error: expected a type specifier
foo.h(3): error: expected a type specifier
foo.h(3): error: explicit type is missing ("int" assumed)
foo.h(3): error: qualifier must be base class of "thrust::detail::Foo<T>"
```

These have the nice advantage of being able to run multiple assertion checks in a single test, but they limit our ability to develop the library. Additionally, the techniques used to implement them aren't compatible with nvc++ and these tests are disabled for that compiler.

We should replace these with compilation tests using CMake. This would also let us test for other expected build passes/failures besides static asserts.

In the meantime, the syntax issue can be worked-around using a raw C++11 `static_assert`.

**Comment by anriltine** [comment]
I’m new here and would like to help move this forward if I can.

### What #807 is asking for

[#807](https://github.com/NVIDIA/cccl/issues/807) describes Thrust tests that used to **redefine `THRUST_STATIC_ASSERT`** so “failed” static assertions could be observed **at runtime**, and asks to replace that with **CMake-driven compile-fail / build-failure tests**. The issue also notes that approach was **awkward for `nvcc` diagnostics** and **problematic with nvc++**, and that **CMake compile tests** could cover **other** expected compile failures too, not only `static_assert`.

### What has already changed on `main`

- **[#3971](https://github.com/NVIDIA/cccl/pull/3971)** (*Deprecate and replace `THRUST_STATIC_ASSERT`*) was **merged 2025-03-03**. It updates `thrust/thrust/detail/static_assert.h` so `THRUST_STATIC_ASSERT` / `THRUST_STATIC_ASSERT_MSG` expand to ordinary **`static_assert`**, and it refactors several Thrust tests and generic code paths that depended on the old macro shape (merge commit `eb178cb39a`, 17 files).
- A repo-wide search shows **no** remaining `runtime_static_assert` machinery under `thrust/`.
- The **specific** `generate_const_iterators.cu` example linked from the original Thrust repo in the issue text is **not** present under current `thrust/testing/` (there are `generate.cu` / `cuda/generate.cu`, but not that filename).

So the **core library / macro half** of the pain in #807 is largely addressed; **#807 itself remains open** on GitHub, so the **testing / CMake** half is still the tracking item.

### What still aligns with the original request

- **Shared infrastructure** already exists: `cccl_add_xfail_compile_target_test` is implemented in [`cmake/CCCLUtilities.cmake`](https://github.com/NVIDIA/cccl/blob/main/cmake/CCCLUtilities.cmake) and is **used in production** today (e.g. **CUB** in [`cub/test/CMakeLists.txt`](https://github.com/NVIDIA/cccl/blob/main/cub/test/CMakeLists.txt)). That matches the direction #807 describes.
- **Thrust’s main** [`thrust/testing/CMakeLists.txt`](https://github.com/NVIDIA/cccl/blob/main/thrust/testing/CMakeLists.txt) currently **does not** register analogous xfail-compile targets there — only the usual unit-test glob and `add_subdirectory(cmake|cpp|cuda|omp)`.
- **Housekeeping:** the same file still declares **`thrust_declare_test_restrictions(unittest_static_assert …)`** even though there is **no** `unittest_static_assert` test source in `thrust/testing/` — that line looks **stale** and is worth cleaning up when someone works this issue.

### nvc++ / CMake context (related, separate)

[#4232](https://github.com/NVIDIA/cccl/issues/4232) (*CMake fails when using nvc++ as a c++ compiler*, Thrust component) was a **closed** report about **CMake configuration** with **nvc++**; it is **not** the same as #807, but it illustrates that **Thrust + nvc++ + CMake** has had sharp edges outside this specific testing technique.

---

**Bottom line:** #807 is still the right place to track **Thrust-side CMake compile-fail coverage** (and small CMake cleanup), while [#3971](https://github.com/NVIDIA/cccl/pull/3971) already landed the **macro / source** side that motivated much of the original write-up.
